src/dma/traits.rs

```diff
@@ -86,80 +98,82 @@ pub trait Stream: StreamISR + crate::Sealed {
+    /// re-enabling the stream. It's also advisable to clear all interrupt flag before re-enabling
+    /// the stream.
+    ///
+    /// # Safety
+    ///
+    /// Disabling the stream before end of transfers may produce invalid data.
+    unsafe fn disable(&mut self);
 
     /// Set the channel for the (chsel) the DMA stream.
-    fn set_channel<const C: u8>(&mut self)
-    where
-        ChannelX<C>: Channel;
+    fn set_channel(&mut self, channel: DmaChannel);
 
     /// Set the priority (pl) the DMA stream.
     fn set_priority(&mut self, priority: config::Priority);
 
+    /// Set the peripheral increment offset (pincos)
+    fn set_peripheral_increment_offset(&mut self, value: PeripheralIncrementOffset);
+
     /// Set the memory size (msize) for the DMA stream.
     ///
     /// # Safety
     /// This must have the same alignment of the buffer used in the transfer.
-    /// Valid values:
-    ///     * 0 -> byte
-    ///     * 1 -> half word
-    ///     * 2 -> word
-    unsafe fn set_memory_size(&mut self, size: u8);
+    unsafe fn set_memory_size(&mut self, size: DmaDataSize);
 
     /// Set the peripheral memory size (psize) for the DMA stream.
     ///
     /// # Safety
     /// This must have the same alignment of the peripheral data used in the transfer.
-    /// Valid values:
-    ///     * 0 -> byte
-    ///     * 1 -> half word
-    ///     * 2 -> word
-    unsafe fn set_peripheral_size(&mut self, size: u8);
+    unsafe fn set_peripheral_size(&mut self, size: DmaDataSize);
 
     /// Enable/disable memory increment (minc) for the DMA stream.
     fn set_memory_increment(&mut self, increment: bool);
 
     /// Enable/disable peripheral increment (pinc) for the DMA stream.
     fn set_peripheral_increment(&mut self, increment: bool);
 
+    /// Enable/disable circular mode (circ) for the DMA stream.
+    fn set_circular_mode(&mut self, value: bool);
+
     /// Set the direction (dir) of the DMA stream.
-    fn set_direction<D: Direction>(&mut self, direction: D);
-
-    /// Convenience method to configure the 4 common interrupts for the DMA stream.
-    fn set_interrupts_enable(
-        &mut self,
-        transfer_complete: bool,
-        half_transfer: bool,
-        transfer_error: bool,
-        direct_mode_error: bool,
-    );
-
-    /// Convenience method to get the value of the 4 common interrupts for the DMA stream.
-    /// The order of the returns are: `transfer_complete`, `half_transfer`, `transfer_error` and
-    /// `direct_mode_error`.
-    fn get_interrupts_enable() -> (bool, bool, bool, bool);
+    fn set_direction(&mut self, direction: DmaDirection);
+
+    /// Set the flow controller (pfctrl).
+    fn set_flow_controller(&mut self, value: DmaFlowController);
+
+    /// Convenience method to configure several interrupts of the DMA stream.
+    ///
+    /// Note: fifo_error interrupt is not concerend because it's in a different register
+    fn set_common_interrupts(&mut self, interrupts: DmaCommonInterrupts);
```

Comment:
for other peripherals we use `listen/unlisten` here.

Reply:
for this one, I suggest this
```Rust
/// Listen/Unlisten several interrupts of the DMA stream.
fn listen_common_interrupts(&mut self, interrupts: DmaCommonInterrupts);
```

for individual interrupts setting i can use listen/unlisten methods

Reply:
If you are using `DmaCommonInterrupts` as structure, yes.

I would use `bitflags` instead of simple structure here.
Advantage is you can do `OR` (`|`) operations on flags.

Difference is in semantics.
When you `clear` the flag other flags are unchanged. Same for `bitflags` + both independent `listen` and `unlisten`.

Current `set_interrupts_enable`/`set_common_interrupts` behavior is different. You need to set/reset all flags, not just that you expect to be changed.

Although I can be wrong. This is breaking change.

And I'm not sure which variant can be better optimized.

Reply:
done, i turned all  "set_smth_interrupt" in to set of "listen_smth and unlisten_smth"